ubo_app/display.py

```diff
@@ -41,51 +40,59 @@
         cs=cs_pin,
         dc=dc_pin,
         rst=reset_pin,
-        baudrate=60000000,
+        baudrate=70000000,
     )
 else:
     display = cast(ST7789, Fake())
 
 
 def render_on_display(
     *,
-    rectangle: tuple[int, int, int, int],
-    data: NDArray[np.uint8],
-    data_hash: int,
-    last_render_thread: Thread,
+    regions: list[Region],
 ) -> None:
     """Transfer data to the display via SPI controller."""
-    data_ = data.astype(np.uint16)
-    color = (
-        ((data_[:, :, 0] & 0xF8) << 8)
-        | ((data_[:, :, 1] & 0xFC) << 3)
-        | (data_[:, :, 2] >> 3)
-    )
-    data_bytes = bytes(
-        np.dstack(((color >> 8) & 0xFF, color & 0xFF)).flatten().tolist(),
-    )
-    if last_render_thread:
-        last_render_thread.join()
-    render_block(rectangle, data_bytes)
-    compressor = zlib.compressobj(wbits=-zlib.MAX_WBITS)
-    store._dispatch(  # noqa: SLF001
-        [
-            DisplayRenderEvent(
-                data=data.tobytes(),
-                data_hash=data_hash,
-                rectangle=rectangle,
+    for region in regions:
+        rectangle = region['rectangle']
+        data = region['data'].astype(np.uint16)
+        color = (
+            ((data[:, :, 0] & 0xF8) << 8)
+            | ((data[:, :, 1] & 0xFC) << 3)
+            | (data[:, :, 2] >> 3)
+        ).copy()
+        data_bytes = (
+            color.astype(np.uint16).view(np.uint8).reshape(-1, 2)[:, ::-1].tobytes()
+        )
+        render_block(
+            (
+                rectangle[1],
+                rectangle[0],
+                rectangle[3] - 1,
+                rectangle[2] - 1,
             ),
-            DisplayCompressedRenderEvent(
-                compressed_data=compressor.compress(data.tobytes())
-                + compressor.flush(),
-                data_hash=data_hash,
-                rectangle=rectangle,
-            ),
-        ],
-    )
-
-
-@store.view(lambda state: state.display.is_paused)
+            data_bytes,
+        )
+        compressor = zlib.compressobj(wbits=-zlib.MAX_WBITS)
```

Comment:
@sassanh this is cool! so you are compressing the data before sending it over the redux bus? I see both `DisplayRenderEvent` and `DisplayCompressedRenderEvent` events; that means you are sending both compressed and non-compressed display data?

Reply:
@mehrdadfeller There is a trade-off here, uncompressed raw data is easier for clients to consume, but takes more bandwidth and grpc-web clients can't use it, compressed data takes less bandwidth and our web client over grpc-web is using it, but it enforces a decompression level for the consumer.

That's why we are providing (and not "sending") both, it doesn't mean that a client receives both and the bandwidth will be populated by both, we only "send" the one the client has subscribed to on each channel.